File: DataStructures/Trie/implementation/Trie.java
// Implementation of Trie data structure in java.

/*
Trie object will be instantiated and called as such:
Trie obj = new Trie();
obj.insert(word);
param_0 = obj.search(word);
param_1 = obj.startsWith(word);
*/

// Create Trie node
class TrieNode {
    public char val;
    public boolean isEnd;
    public TrieNode[] children;

    // Constructor
    public TrieNode(char val) {
        this.val = val;
        this.isEnd = false;
        this.children = new TrieNode[26];
    }
}

class Trie {
    private TrieNode root;

    // Create root node
    public Trie() {
        root = new TrieNode(' ');
    }

    // Insert a word into the Trie
    public void insert(String word) {
        TrieNode node = root;

        for (char c : word.toCharArray()) {
            int index = c - 'a';
            if (node.children[index] == null) {
                node.children[index] = new TrieNode(c);
            }
            node = node.children[index];
        }
        node.isEnd = true;  // Mark the end of a word
    }

    // Search for a word in the Trie
    public boolean search(String word) {
        TrieNode node = root;

        for (char c : word.toCharArray()) {
            int index = c - 'a';
            if (node.children[index] == null) {
                return false;
            }
            node = node.children[index];
        }

        return node.isEnd;
    }

    // Check if a word in Trie starts with prefix
    public boolean startsWith(String prefix) {
        TrieNode node = root;

        for (char c : prefix.toCharArray()) {
            int index = c - 'a';
            if (node.children[index] == null) {
                return false;
            }
            node = node.children[index];
        }

        return true;
    }
}

File: DataStructures/Trie/implementation/Trie.py
# Implementation of Trie data structure in python.

# Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_0 = obj.search(word)
# param_1 = obj.startsWith(prefix)

class TrieNode:
    def __init__(self, val='', is_end=False):
        ''' Create a Trie node '''
        self.val = val
        self.children = {}
        self.is_end = is_end

class Trie:
    def __init__(self):
        ''' Create the root node '''
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        ''' Insert word in Trie '''
        node = self.root

        for c in word:
            if c not in node.children:
                node.children[c] = TrieNode(c)
            node = node.children[c]
        
        node.is_end = True

    def search(self, word: str) -> bool:
        ''' Search for word in Trie '''
        node = self.root

        for c in word:
            if c not in node.children: return False
            node = node.children[c]
        
        return node.is_end

    def startsWith(self, prefix: str) -> bool:
        ''' Check if a word in Trie starts with prefix '''
        node = self.root

        for c in prefix:
            if c not in node.children: return False
            node = node.children[c]
        
        return True

